feat(curriculum): add real-time counter lab (#55838)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-real-time-counter/66bb3e20d3dc5b6d0a21f5dd.md

10. Create a condition that checks if the current character count is greater than `50`. If it is, the user shouldn't be able to enter more characters. 
11. Get the character count display element by its ID and update the text to show the current character count using `innerHTML`.
12. If the character count is equal to `50`, the text should be displayed in `red`.  -->

**Note:** Be sure to link your stylesheet and the JavaScript file in your HTML.

# --hints--

There should be an `h1` element with the text of your topic.

```js
assert.isAbove(document.querySelector('h1')?.innerText.length, 0);
```

You should have a `textarea` element with the `id` of `text-input`. 

```js
assert($('textarea').attr('id') == 'text-input');
```

You should have a `p` element with the `id` of `char-count`. 

```js
assert($('p').attr('id') == 'char-count');
```

The `#char-count` `p` element should initially contain the text `Character Count: 0/50`.

```js
assert(document.querySelector('p')?.innerText, 'Character Count: 0/50');
```

When the `#text-input` element contains the text `hello coder` the `#char-count` element should contain the text `"Character Count: 11/50"`

```js

> const textInput = document.getElementById('text-input');
> const charCount = document.getElementById('char-count');
> 
> textInput.dispatchEvent(new Event('change'));
> 
> assert.strictEqual(charCount.innerText.trim(), 'Character Count: 11/50');
> console.log("current count" + charCount.innerText);

Reviewer: ```suggestion
const textInput = document.getElementById('text-input');
const charCount = document.getElementById('char-count');

textInput.value = "hello coder";
textInput.dispatchEvent(new Event('input'));
textInput.dispatchEvent(new Event('change'));

assert.strictEqual(charCount.innerText.trim(), 'Character Count: 11/50');
```

We need to fire the "input" event too, which is what the solution code listens for. Also, neither of these will really help if we don't change the *value* of the text input first. :3

Author: This test is passing even if nothing is entered in the text area.

Author: I was expecting something different, this test works!

```js
const charCount = document.querySelector('p');
assert.equal(charCount.id, 'char-count');
```

The `#char-count` `p` element should initially contain the text `Character Count: 0/50`.

```js
assert(document.querySelector('p')?.innerText, 'Character Count: 0/50');
```

When the `#text-input` element contains the text `hello coder` the `#char-count` element should contain the text `"Character Count: 11/50"`

```js
const textInput = document.getElementById('text-input');
const charCount = document.getElementById('char-count');
textInput.value = "hello coder";
textInput.dispatchEvent(new Event('input'));
textInput.dispatchEvent(new Event('change'));
assert.strictEqual(charCount.innerText.trim(), 'Character Count: 11/50');
```

When the character count is `50`, the text should be displayed in `red`.

```js
const textInput = document.getElementById('text-input');
const charCount = document.getElementById('char-count');
textInput.value = "I am learning a new language and it is called c++.";
textInput.dispatchEvent(new Event('input'));
textInput.dispatchEvent(new Event('change'));
assert.strictEqual(charCount.style.color, 'red');
```

> If character count is greater than `50`, the user shouldn't be able to enter more characters.

Reviewer 2: ```suggestion
If character count is greater than or equal to `50`, the user shouldn't be able to enter more characters.
```

```js
let textInput = document.getElementById("text-input");

> let charCount = textInput.value.length;
> textInput.value = "I am learning a new language and it is called c++.";
> textInput.dispatchEvent(new Event('input'));
> textInput.dispatchEvent(new Event('change'));
> assert.strictEqual(charCount, 50);

Reviewer 2: ```suggestion
textInput.value = "I am learning a new programming language and it is called c++.";
textInput.dispatchEvent(new Event('input'));
textInput.dispatchEvent(new Event('change'));
assert.strictEqual(textInput.value.length, 50);
```

This test wasn't working quite right. The string you had was exactly 50 characters, we want to make sure that a string longer than 50 isn't allowed.

---
id: 66bb3e20d3dc5b6d0a21f5dd
title: Build a Real Time Counter
challengeType: 14
dashedName: build-a-real-time-counter
demoType: onClick
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. **Do not copy this demo project**.

**User Stories:**

1. You should have an `h1` element with the text `Real-Time Character Counter`.

> 1. You should have a `textarea` element with the `id` of `textInput`. It should have a `placeholder` attribute with the text `Type something...`.

Reviewer 3: ```suggestion
1. You should have a `textarea` element with the `id` of `text-input`. It should have a `placeholder` attribute with the text `Type something...`.
```
I would use `text-input` in place of `textInput`

> 1. There should be a `p` element with the `id` of `charCount`. It should initially contain the text `Character Count: 0/50`. This placeholder text should be replaced with the current count of characters in the `textarea` element.

Reviewer 3: ```suggestion
1. There should be a `p` element with the `id` of `char-count`. It should initially contain the text `Character Count: 0/50`. This placeholder text should be replaced with the current count of characters in the `textarea` element.
```
I would use `char-count` instead.

> 1. You should attach an event listener to the `textInput` element that listens for the `input` event. 

Reviewer 3: ```suggestion
1. You should attach an event listener to the `text-input` element that listens for the `input` event. 
```

1. Inside the event trigger function, you should get the value of the `textarea` element and store it in the variable `textInput`. You should also declare a variable `charCount` and set it equal to the length of the `textInput` variable.
1. Create a condition that checks if the current character count is greater than `50`. If it is, the user shouldn't be able to enter more characters. 
1. Get the character count display element by its ID and update the text to show the current character count using `innerHTML`.
1. If the character count is equal to `50`, the text should be displayed in `red`. 

# --hints--

There should be an `h1` element with the text of your topic.

```js
assert.isAbove(document.querySelector('h1')?.innerText.length, 0);
```

> You should have a `textarea` element with the `id` of `textInput`. 

Reviewer 3: ```suggestion
You should have a `textarea` element with the `id` of `text-input`. 
```

```js

> assert($('textarea').attr('id') == 'textInput');

Reviewer 3: ```suggestion
assert($('textarea').attr('id') == 'text-input');
```

```

Your `textarea` should have a `placeholder` attribute with the text `Type something...`.
    
```js
assert($('textarea').attr('placeholder') == 'Type something...');
```

> You should have a `p` element with the `id` of `charCount`. 

Reviewer 3: ```suggestion
You should have a `p` element with the `id` of `char-count`. 
```
Following the previous comment.

```js

> assert($('p').attr('id') == 'charCount');

Reviewer 3: ```suggestion
assert($('p').attr('id') == 'char-count');
```

```

> The `#id` `p` element should initially contain the text `Character Count: 0/50`.

Reviewer 3: ```suggestion
The `#char-count` `p` element should initially contain the text `Character Count: 0/50`.
```

```js
assert(document.querySelector('p')?.innerText, 'Character Count: 0/50');
```

> You should attach an event listener to the `textInput` element.

Reviewer 3: ```suggestion
You should attach an event listener to the `text-input` element.
```

```js

> assert.match(code, /document\.getElementById\(\"textInput\"\)\.addEventListener/);

Reviewer 3: As Jessica mentioned, one can use `querySelector` instead of `getElementById`. But also, you can get the element and store it in a variable and have `<variableName>.addEventListener`

```

> Your event listener on `#textInput` should listen for the `input` event.

Reviewer 3: ```suggestion
Your event listener on `#text-input` should listen for the `input` event.
```

```js
assert.match(code, /document\.getElementById\(\"textInput\"\)\.addEventListener\(\"input\"/);
```

Inside the event trigger function, you should get the value of the `textarea` element using its id and store it in the variable `textInput`. 

```js
assert.match(code, /document\.getElementById\("textInput"\)\.addEventListener\("input",\s*function\s*\(\)\s*{\s+let\s+textInput\s*=\s*document\.getElementById\("textInput"\);/);
```

Inside the event trigger function, you should declare a variable `charCount` and set it equal to the length of the `textInput` variable.

```js
assert.match(code, /document\.getElementById\("textInput"\)\.addEventListener\("input",\s*function\s*\(\)\s*{\s+let\s+textInput\s*=\s*document\.getElementById\("textInput"\);\s+let\s+charCount\s*=\s*textInput\.value\.length;/);
```

> xoxo Inside the event trigger function, you should create a condition that checks if the current character count is greater than `50`. 

Reviewer 3: ```suggestion
Inside the event trigger function, you should create a condition that checks if the current character count is greater than `50`. 
```

```

Inside the event trigger function, if the character count is greater than `50`, you should use `substring` operation to limit the number of characters to `50` so the user can't enter more characters. Also, assign the substring to the `textInput` variable.

```js
assert.match(code, /document\.getElementById\("textInput"\)\.addEventListener\("input",\s*function\s*\(\)\s*{\s+let\s+textInput\s*=\s*document\.getElementById\("textInput"\);\s+let\s+charCount\s*=\s*textInput\.value\.length;\s+if\s*\(charCount\s*>\s*50\)\s*\{\s+textInput\.value\s*=\s*textInput\.value\.substring\(0,\s*50\);/);
```

If the character count is greater than `50`, the value of `charCount` should be set to `50`.

```js
assert.match(code, /document\.getElementById\("textInput"\)\.addEventListener\("input",\s*function\s*\(\)\s*{\s+let\s+textInput\s*=\s*document\.getElementById\("textInput"\);\s+let\s+charCount\s*=\s*textInput\.value\.length;\s+if\s*\(charCount\s*>\s*50\)\s*\{\s+textInput\.value\s*=\s*textInput\.value\.substring\(0,\s*50\);\s+charCount\s*=\s*50;/);
```

> You should select the character count display element by its ID `charCount` and store it in the variable `charCountDisplay`.

Reviewer 3: ```suggestion
You should select the character count display element by its id `char-count` and store it in the variable `charCountDisplay`.
```

```js
assert.match(code, /document\.getElementById\("textInput"\)\.addEventListener\("input",\s*function\s*\(\)\s*{\s+let\s+textInput\s*=\s*document\.getElementById\("textInput"\);\s+let\s+charCount\s*=\s*textInput\.value\.length;\s+if\s*\(charCount\s*>\s*50\)\s*\{\s+textInput\.value\s*=\s*textInput\.value\.substring\(0,\s*50\);\s+charCount\s*=\s*50;\s*}\s+let\s+charCountDisplay\s*=\s*document\.getElementById\("charCount"\);/);
```

> You should update the `#charCount` text to show the current character count using `innerHTML` and using the string template `Character Count: ${charCount}/50`.

Reviewer 3: ```suggestion
You should update the `#char-count` text to show the current character count using `innerHTML` and using the string template `Character Count: ${charCount}/50`.
```

```

If the character count is equal to `50`, the text should be displayed in `red`.

```js
assert.match(code, /document\.getElementById\("textInput"\)\.addEventListener\("input",\s*function\s*\(\)\s*{\s+let\s+textInput\s*=\s*document\.getElementById\("textInput"\);\s+let\s+charCount\s*=\s*textInput\.value\.length;\s+if\s*\(charCount\s*>\s*50\)\s*\{\s+textInput\.value\s*=\s*textInput\.value\.substring\(0,\s*50\);\s+charCount\s*=\s*50;\s*}\s+let\s+charCountDisplay\s*=\s*document\.getElementById\("charCount"\);\s+charCountDisplay\.innerHTML\s*=\s*`Character Count:\s*\${charCount}\/50`;\s+if\s*\(charCount\s*===\s*50\)\s*\{\s+charCountDisplay\.style\.color\s*=\s*"red";\s*}/);
```

# --seed--

## --seed-contents--

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Real Time Counter</title>

>     <link rel="stylesheet" href="styles.css">

Reviewer 3: ```suggestion
```
As we decided to let campers write the link and the script themselves

</head>

<body>

</body>

> <script src="script.js"></script>

Reviewer 3: ```suggestion
```

</html>
```

```css

```

```js
    
```

# --solutions--

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Real-time Character Counter</title>
    <link rel="stylesheet" href="styles.css">

<body>
    <div class="container">
        <h1>Real-Time Character Counter</h1>

>         <textarea id="textInput" placeholder="Type something..."></textarea>

Reviewer 3: ```suggestion
        <textarea id="text-input" placeholder="Type something..."></textarea>
```

>         <p id="charCount">Character Count: 0/50</p>

Reviewer 3: ```suggestion
        <p id="char-count">Character Count: 0/50</p>
```

    </div>

> </body>
> <script src="script.js"></script>

Reviewer 3: ```suggestion
<script src="script.js"></script>
</body>
```

</html>
```

```css
body {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100vh;
    font-family: Arial, sans-serif;
    background-color: #c0c1ec;
    margin: 0;
}

h1{
    background-color: #f2f48b;
    padding-left: 10px;
    padding-right: 10px;
}
.container {
    text-align: center;
}

textarea {
    width: 300px;
    height: 100px;
    font-size: 18px;
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 5px;
    resize: none;
    display: block;
    margin: 0 auto;
    border: 1px solid slategray;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
}

textarea:focus {
    outline: none;
    border-color: #f2f48b;
}

> #charCount {

Reviewer 3: ```suggestion
#char-count {
```

    margin-top: 10px;
    font-size: 18px;
    color: darkslategray;
    font-weight: bold;
}
```

```js
document.getElementById("textInput").addEventListener("input", function () {
  let textInput = document.getElementById("textInput");
  let charCount = textInput.value.length;

  if (charCount > 50) {
    textInput.value = textInput.value.substring(0, 50);
    charCount = 50;
  }

>   let charCountDisplay = document.getElementById("charCount");

Reviewer 3: ```suggestion
  let charCountDisplay = document.getElementById("char-count");
```
